### DISCUSSION TITLE
Solr index get's incorrect 20% of the time, when deleting value from two properties with different cardinality

### DISCUSSION THREAD

**Discussion by urblin33** [JanusGraph/janusgraph#2422] [Q&A]
I'm new to this forum, but I'm pretty sure this is an existing bug in version 0.5.3, running with solr 7.3.3.
We have experienced this in our production environment with older versions, so I first verified that it still existed using the newest stable version. 
Should I go ahead and create an issue?
If I in one transaction on a vertex delete values in a property with cardinality SET and a property with cardinality SINGLE, I experience that the property with cardinality SET, get's incorrect/wiped out from the solr index 20% of the time. The stored property is correct. I made me a test to verify.   and property index.search.solr.wait-searcher is set to true.

    @Test
    public void removePropertyInList(){
        int foundInSolr=0;
        for (int i=0;i<50;i++) {
            String[] bylineIds = {"3", "1","2"};
            Graph g = dbAdapter.getGraph();

            //Create vertex and add some properties
            Vertex article = g.addVertex("Article");
            article.property("name","kalle");
            article.property("typeId","204");
            for (String s : bylineIds) {
                article.property("bylineIds", s);
            }
            g.tx().commit();

            // remove all but one value on bylineIds parameter with Cardinality SET
            article.properties("bylineIds").forEachRemaining(v -> {
                if (!v.value().equals("2")){
                    v.remove();
                }
            });

            //the test succeed when using an extra commit
            //g.tx().commit();

            // then removed another property, the test succeeds if we do not remove another property
            article.property("name").remove();

            g.tx().commit();

            // look for it through solr index and make sure we find the one we just created
            JanusGraph jg = (JanusGraph) g;
            boolean foundV = jg.indexQuery("vertices", "v.bylineIds:(2)").vertexStream().filter(v->v.getElement().id().equals(article.id())).count() ==1;
            if (foundV){
                foundInSolr++;
            }
        }
        assertEquals(50,foundInSolr);

    }

**Comment by li-boxuan** [comment]
~~I am not able to reproduce it: https://github.com/li-boxuan/janusgraph/commit/6239118396ac4afeb1351225299e92665e2431fa my test code is not 100% same as yours, though.~~

Can you provide complete java/groovy code that can reproduce? Your code does not include how properties are indexed.

**Reply by urblin33** [reply]
Was hoping to avoid that :) but will do, Thank you for trying!

**Reply by urblin33** [reply]
package com.mmab.soldr.api.swagger.impl;

import org.apache.tinkerpop.gremlin.structure.Vertex;
import org.janusgraph.core.Cardinality;
import org.janusgraph.core.JanusGraph;
import org.janusgraph.core.JanusGraphFactory;
import org.janusgraph.core.JanusGraphVertex;
import org.janusgraph.core.schema.JanusGraphManagement;
import org.janusgraph.core.util.JanusGraphCleanup;
import org.janusgraph.diskstorage.BackendException;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class VertexApiServiceImplTest  {

    public static final String ARTICLE = "Article";
    public static final String OTHER = "other";
    public static final String EXTRA = "extra";
    public static final String SOME_IDS = "someIds";

    private static JanusGraphFactory.Builder createBuilder() {

        JanusGraphFactory.Builder builder = JanusGraphFactory.build();

        // builder.set("storage.backend", "inmemory");
        builder.set("storage.backend", "cassandra");
        builder.set("storage.hostname", "127.0.0.1");
        builder.set("storage.username", "cassandra");
        builder.set("storage.password", "cassandra");

        builder.set("storage.cassandra.keyspace", "titan");

        builder.set("index.search.backend", "solr");
        builder.set("index.search.solr.mode", "http");
        builder.set("index.search.solr.http-urls", "http://localhost:8983/solr");

        // Make sure we wait for SOLR indexes to be committed and all new
        // searchers will see the changes
        builder.set("index.search.solr.wait-searcher", "true");

        // this option might be good to set to ensure we have enough indexes
        builder.set("query.force-index", "true");

        builder.set("ids.store-name", "titan_ids");

        return builder;
    }


    private void setupSchema() throws BackendException {
        // code for clearing graph
        JanusGraph g =createBuilder().open();
        JanusGraphCleanup.clear(g);
        g.close();
        g =createBuilder().open();
        // clear solr
        //g.close();
        //SolrClearStorageAdapter solrService = new SolrClearStorageAdapter();
        //solrService.clearStorage();

        JanusGraphManagement graphManagement = g.openManagement();
        graphManagement.makeVertexLabel(ARTICLE).make();
        graphManagement.makePropertyKey(OTHER).dataType(String.class).cardinality(Cardinality.SINGLE).make();
        graphManagement.makePropertyKey(EXTRA).dataType(String.class).cardinality(Cardinality.SINGLE).make();
        graphManagement.makePropertyKey(SOME_IDS).dataType(String.class).cardinality(Cardinality.SET).make();

        JanusGraphManagement.IndexBuilder verticesIndexBuilder = graphManagement.buildIndex("vertices", JanusGraphVertex.class);
        verticesIndexBuilder.addKey(graphManagement.getPropertyKey(OTHER));
        verticesIndexBuilder.addKey(graphManagement.getPropertyKey(EXTRA));
        verticesIndexBuilder.addKey(graphManagement.getPropertyKey(SOME_IDS));
        verticesIndexBuilder.buildMixedIndex("search");

        graphManagement.commit();
        g.close();
    }


    @Test
    @Disabled
    public void removePropertyInList() throws BackendException {

        setupSchema();

        JanusGraph g =createBuilder().open();



        int foundInSolr=0;
        for (int i=0;i<50;i++) {
            String[] someIds = {"3", "1","2"};


            //Create vertex
            Vertex article = g.addVertex(ARTICLE);

            article.property(OTHER,"kalle");
            article.property(EXTRA,"204");
            for (String s : someIds) {
                article.property(SOME_IDS, s);
            }
            g.tx().commit();

            // remove all but one value on someIds parameter with Cardinality SET
            article.properties(SOME_IDS).forEachRemaining(v -> {
                if (!v.value().equals("2")){
                    v.remove();
                }
            });

            // works correct when using an extra commmit
            //g.tx().commit();

            // the remove another property, works correct if we do not remove another property
            article.property(OTHER).remove();


            g.tx().commit();

            // count nr of id's
            final AtomicInteger count=new AtomicInteger(0);
            article.properties(SOME_IDS).forEachRemaining(p->count.incrementAndGet());
            // verify found in database
            assertEquals(1,count.get());

            // look for it through solr index and make sure we find the one we just created

            boolean foundV = g.indexQuery("vertices", "v.someIds:(2)").vertexStream().filter(v->v.getElement().id().equals(article.id())).count() ==1;
            if (foundV){
                foundInSolr++;
            }

            //It's correct in the database, but when running this 50 times and having a look in solr,
            // we find the property someIds is empty in about 20% of the cases
            // If we do not remove "another" property, it works, or if we do an extra commit,
            // between the removes it also works as expected

        }
        System.out.println("solrCount "+foundInSolr);
        assertEquals(50,foundInSolr);

    }



}

**Reply by urblin33** [reply]
hope the code above is easier to use for reproducing, sorry for being lazy, a little stressed by some production issues.

**Reply by li-boxuan** [reply]
Thanks, I reproduced it in https://github.com/li-boxuan/janusgraph/tree/bug/delete-set-value
Only Solr test fails, see https://github.com/li-boxuan/janusgraph/runs/1836798839, so it seems to be a Solr specific issue
Could you create a bug on https://github.com/JanusGraph/janusgraph/issues?

**Reply by urblin33** [reply]
Created issue
https://github.com/JanusGraph/janusgraph/issues/2434
